rdm/0007-access-control.md

```diff
@@ -0,0 +1,190 @@
+- Delete: Ability to delete a record.
+- List: Ability to perform searches (note, for a record to appear in the search result it is subject to the above mentioned read action).
+
+Records linked with files (possible for all records) have the additional following actions.
+- Create files: Ability to create files associated with a record. 
+- Read files: Ability to read files associated with a record.
+- Update files: Ability to update files associated with a record.
+- Delete files: Ability to delete files associated with a record.
+- Administrate files: Ability to perform all possible actions on files associated with a record, including reading previous versions of files.
+
+Deposit records have the additional following actions:
+- Publish: Ability to publish a new record.
+- Edit: Ability to edit an already published record. This differs from the record update action.
+- New version: Ability to create a new version of an existing record.
+
+It is possible to extend this model with more actions for each data model as needed.
+
+**Requiring permissions**
+
+A record must store permissions inside the metadata. This can be as simple as a list of owners, or very complex based on custom rules (e.g. books can only be curated by books curators). This rules are translated into `generators`.
+
+**Generators**
+
+A *generator* implements a rule to allow or deny the access to perform an action.  
+
+The generator would return a set of *needs* and *excludes* for each action. The *needs* are the constraints the user has to meet in order to be allowed to perform an action (e.g. be the owner of the record). The *excludes* specify a set of blacklisting constraints, which have precedence over the *needs*. 
+
+For example, the *record owners* generator would look like:
+
+``` python
+Document:
+    _owners: [1, 2, 3]
+
+RecordOwner:
+    needs:
+        return Document._owners
+    excludes:
+        pass
+```
+
+Since the record owner does not set any specific blacklist (*excludes*), the generator would result in an access granted for users with id 1, 2 or 3.
+
+In addition, generators return a set of *query filters*, which are used only when searching. The set of rules returned there are in the form of an Elasticsearch Query object. The `RecordOwner` generator would therefore look like:
+
+``` python
+RecordOwner:
+    needs:
+        return Document._owners
+    excludes:
+        pass
+    query_filter:
+        return Query(owner=user.id)
+```
+
+As a result, it can be seen that the *query filter* inverts the paradigm with respect to the *needs*/*excludes*. Instead of receiving a record and returning its owners, it takes the id of the user performing the action and returns a query object. This is due to the nature of the *list* (search) operation. The check must not be *Can user A access record R, for each one of the records in the system*. This would not be scalable. Therefore, the check must be *which documents can be accessed by the user A*.
```

Comment:
nit: I would re-phrase the second part `This is due to...` to explain it a little bit better.